### PR TITLE
Refactor weak-reference handling code in SGen

### PR DESCRIPTION
**Pull request by evincarofautumn**
There were a few tables of GC handle data in `gc.c`, whose information was being duplicated by the hash tables in `sgen-fin-weak-hash.c`. Removing these hash tables and using the handle table directly is both simpler and considerably more efficient. In addition, any operation on the GC handles tables used to take a lock. Now the implementation is lock-free.

To do:

 * Reinstate the segregation of weak links by minor vs. major heap

CC @schani

### CODE REVIEW

**Comment by evincarofautumn** [comment]
Rebased, cleaned up the history, and added some missing code for nulling links in domain unloading. I think this is finally good to go.

**Comment by schani** [comment]
This PR introduces a crash in Roslyn.  This is how I run it:

    ~/W/m/b/t/r/corlib (master)$ while env MONO_GC_PARAMS=nursery-size=1m,major=marksweep ./run.sh
                                     true
                                 end

It takes many attempts to crash.  The commit this PR is based on does not crash, even after about 1.5 hours.

**Comment by schani** [comment]
Also, Roslyn runs quite a bit slower with the lock-free weak links, and it's mostly not due to increased GC time.  Old:

    Minor GC collections                : 586
    Major GC collections                : 20
    Minor GC time                       : 2324.14 ms
    Major GC time                       : 2239.42 ms
    Major GC time concurrent            : 0.00 ms
    
    real	0m10.857s
    user	0m23.737s
    sys	0m1.718s

New:

    Minor GC collections                : 584
    Major GC collections                : 20
    Minor GC time                       : 2585.61 ms
    Major GC time                       : 2245.42 ms
    Major GC time concurrent            : 0.00 ms

    real	0m13.121s
    user	0m29.497s
    sys	0m1.673s

**Comment by evincarofautumn** [comment]
This is working now. [Benchmark results](https://drive.google.com/a/xamarin.com/file/d/0B9vUy441CSjgOGhubHQyT2dWRUk/view?usp=sharing) show no difference for most benchmarks. GC times are a few percent longer on F# & IronJS, which both use weak references. Looking into why.

**Comment by evincarofautumn** [comment]
The benchmarks that were slower were using monitors. These now use GC handles, and since they were never being freed, the GC handles table grew quite large and became costly to iterate during a GC. Freeing the monitors solves that issue, and brings performance back on par with master.

File: mono/metadata/monitor.c [hunk 1]
```diff
@@ -475,7 +475,7 @@ mono_monitor_try_enter_internal (MonoObject *obj, guint32 ms, gboolean allow_int
 		mono_monitor_allocator_lock ();
 		mon = mon_new (id);
 		if (InterlockedCompareExchangePointer ((gpointer*)&obj->synchronisation, mon, NULL) == NULL) {
-			mono_gc_weak_link_add (&mon->data, obj, TRUE);
+			mon->data = (void *)(guint64)mono_gchandle_new_weakref (obj, TRUE);
```
Comment (kumpera) [inline mono/metadata/monitor.c]: Casting to size_t has the same effect and it not a warning on 32bits.

File: mono/sgen/sgen-gc.h [hunk 2]
```diff
@@ -942,6 +932,17 @@ sgen_is_object_alive_for_current_gen (char *object)
...
+/* GC handles */
+
+void sgen_init_gchandles (void);
+
+void sgen_null_links_if (SgenObjectPredicateFunc predicate, void *data, int generation, gboolean track);
+
+void sgen_gchandle_set_target (guint32 gchandle, GCObject *obj);
+void sgen_mark_normal_gc_handles (void *addr, SgenUserMarkFunc mark_func, void *gc_data);
+gpointer sgen_gchandle_get_metadata (guint32 gchandle);
+void sgen_gchandle_iterate (GCHandleType handle_type, int max_generation, gpointer callback(gpointer, GCHandleType, int, gpointer), gpointer user);
```
Comment (kumpera) [inline mono/sgen/sgen-gc.h]: function pointer should have a typedef of their own or, at the very least, argument names bound

File: mono/metadata/sgen-bridge.c [hunk 3]
```diff
@@ -179,9 +179,12 @@ null_weak_links_to_dead_objects (SgenBridgeProcessor *processor, int generation)
 	}
 
 	/* Null weak links to dead objects. */
-	sgen_null_links_if (is_bridge_object_dead, &alive_hash, GENERATION_NURSERY);
-	if (generation == GENERATION_OLD)
-		sgen_null_links_if (is_bridge_object_dead, &alive_hash, GENERATION_OLD);
+	sgen_null_links_if (is_bridge_object_dead, &alive_hash, GENERATION_NURSERY, FALSE);
```
Comment (kumpera) [inline mono/metadata/sgen-bridge.c]: This looks like something that could be modified to be single pass.

**Comment by kumpera** [comment]
I found a few issues:

It looks like the new behavior for unloading a domain is to free all gchandles pointing to it.
This is, AFAICT, a breaking change from the existing behavior.

It's not clear if we preserved the behavior of mono_gchandle_set_target with a NULL argument.

Given this adds a lock free implementation of the gc handle table, it would be great if it added a
threaded test to it to the GC test suite.

How much testing/stressing has this received? We're a few weeks away from branching 4.2, so
we might not get enough testing of this feature before the release.

File: mono/metadata/monitor.c (hunk 1 above)
Comment (evincarofautumn) [inline mono/metadata/monitor.c]: Good catch. Should technically be `uintptr_t`, but we already assume `sizeof (size_t) == sizeof (void *)` all over the place, so…eh.

**Comment by evincarofautumn** [comment]
> It's not clear if we preserved the behavior of mono_gchandle_set_target with a NULL argument.

The original code had the same behaviour—since there’s no way to obtain a domain from a null pointer, it would add the current domain to the domain IDs array. There was a comment:

```
/* FIXME, what to use when obj == null? */
```

Somehow it got lost. I can add it back if it’s a concern.

**Comment by schani** [comment]
I don't think we have any obligation to support code that depends on GC handles of a domain staying around after that domain is gone.  We'd have to keep the dangling domain pointer around, for one, and it's insane in any case.

**Comment by evincarofautumn** [comment]
The `sgen-weakref-stress` test seems reasonably stressful. It allocates many ephemeral weak references from several threads while doing garbage collections.

```
time ./mono/mini/mono-sgen ./mono/tests/sgen-weakref-stress.exe
```

`master`:

```
real	0m1.76s
user	0m1.27s
sys	0m0.67s
```

`weak-refs`:

```
real	0m0.32s
user	0m0.36s
sys	0m0.04s
```

Roslyn uses GC handles in monitors:

`master`:

```
real	0m6.707s
user	0m15.925s
sys	0m1.434s
```

`weak-refs`:

```
real	0m6.484s
user	0m15.731s
sys	0m1.134s
```

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
build

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#Nlh4MgbhN1
buildbot logs: http://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo/builds/17
`<botmode>`

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
builddeb

**Comment by schani** [comment]
Overall, performance is not hurt by this change:

  http://xamarin.github.io/benchmarker/front-end/pullrequest.html#B1Yi2xmnHH

**Comment by BrzVlad** [comment]
This improves performance of WeakReference.IsAlive (that we've received complaints about) by 1.5-2 times on serial micro-benchmark scenarios (http://pastebin.com/tGNxN3pL) and up to 40 times on parallel micro-benchmark scenarios (http://pastebin.com/BFXBL8L4)

**Comment by kumpera** [comment]
We added a mono_gc_weak_link_register API. It's not used anywhere, is that on purpose?

**Comment by evincarofautumn** [comment]
@kumpera: That API should have been removed. It looks like I just missed a couple locations. Will PR a fix.

**Pull request merged**